scipy/sparse/_csr.py

```diff
@@ -122,6 +122,59 @@ def tobsr(self, blocksize=None, copy=True):
 
     tobsr.__doc__ = _spbase.tobsr.__doc__
 
+
+    def _broadcast_to(self, shape, copy=False):        
+        if self.shape == shape:
+            return self.copy() if copy else self
+        
+        shape = check_shape(shape, allow_nd=(self._allow_nd))
+
+        if np.broadcast_shapes(self.shape, shape) != shape:
+            raise ValueError("cannot be broadcast")
+        
+        if len(self.shape) == 1 and len(shape) == 1:
+            self.sum_duplicates()
+            if self.nnz == 0: # array has no non zero elements
+                return self.__class__(shape, dtype=self.dtype, copy=False)
+            
+            N = shape[0]
+            data = np.full(N, self.data[0])
+            indices = np.arange(0,N)
+            indptr = np.array([0, N])
+            return self.__class__((data, indices, indptr), shape=shape, copy=False)
+
+        # treat 1D as a 2D row
+        old_shape = self._shape_as_2d
+            
+        if len(shape) != 2:
+            ndim = len(shape)
+            raise ValueError(f'CSR broadcast_to cannot have shape >2D. Got {ndim}D')
+        
+        if self.nnz == 0: # array has no non zero elements
+            return self.__class__(shape, dtype=self.dtype, copy=False)
+        
+        self.sum_duplicates()
+        if all(s == 1 for s in old_shape):
+            # Broadcast a single element to the entire shape
+            data = np.full(shape[0] * shape[1], self.data[0])
+            indices = np.tile(np.arange(shape[1]), shape[0])
+            indptr = np.arange(0, len(data) + 1, shape[1])
+        
+        elif old_shape[0] == 1 and old_shape[1] == shape[1]:
+            # Broadcast row-wise
+            data = np.tile(self.data, shape[0])
+            indices = np.tile(self.indices, shape[0])
+            indptr = np.arange(0, len(data) + 1, len(self.data))
+
+        elif old_shape[1] == 1 and old_shape[0] == shape[0]:
+            # Broadcast column-wise
+            data = np.repeat(self.data, shape[1])
+            indices = np.tile(np.arange(shape[1]), len(self.data))
+            indptr = self.indptr * shape[1]
+
+        return self.__class__((data, indices, indptr), shape=shape, copy=False)
```

Comment:
The idea here is to swap shape, and the concept of row, column when workign with indices and indptr.
So the code the handled rows for CSR will be handling columns for CSC and vice versa.
I made the shapes `M, N` and `oM,oN` so I can easily think about code in terms of CSR (M as rows and N as cols) and I let swap handle the CSC case. And I don't have to change the actual `shape` or `old_shape` to make the code work. Leave them as the true shapes to be used at the end when constructing the result. I'm really only changing `shape[...]` to `M` or `N`, and `old_shape[...]` to 'oM` and `oN`. And I adjust the comments for CSC.
```suggestion
        self.sum_duplicates()
        M, N = self._swap(shape)
        oM, oN = self._swap(old_shape)
        if all(s == 1 for s in old_shape):
            # Broadcast a single element to the entire shape
            data = np.full(M * N, self.data[0])
            indices = np.tile(np.arange(N), M)
            indptr = np.arange(0, len(data) + 1, N)

        elif oM == 1 and oN == N:
            # Broadcast row-wise (columns for CSC)
            data = np.tile(self.data, M)
            indices = np.tile(self.indices, M)
            indptr = np.arange(0, len(data) + 1, len(self.data))

        elif oN == 1 and oM == M:
            # Broadcast column-wise (rows for CSC)
            data = np.repeat(self.data, N)
            indices = np.tile(np.arange(N), len(self.data))
            indptr = self.indptr * N

        return self.__class__((data, indices, indptr), shape=shape, copy=False)
```
Of course these changes need to be followed by moving this function to `_compressed.py` so CSC inherits it. AND `_bsr.py` needs to overwrite `_broadcast_to` to use `_spbase._broadcast_to` instead of the version in `_cs_matrix`.

Reply:
Thanks a lot! I've made these changes at https://github.com/scipy/scipy/pull/21613/commits/050a465b7ffc84c9851935249d82b213ed8e2ab8 and I've added tests to `_csc.py` too! Is it okay to have separate tests in `_csc.py` and `_csr.py`? (since having them in `_base.py` caused some int32 issues in one of the dtype tests because we had sparse array objects (along with matrix) in those tests.)

I also cherry-picked this commit to the CSR broadcasting PR, so if it's easier you can review them there, and then once that's merged I'll just rebase this one.

Reply:
> Is it okay to have separate tests in _csc.py and _csr.py? (since having them in _base.py caused some int32 issues in one of the dtype tests because we had sparse array objects (along with matrix) in those tests.)
Can you help me understand this? Are you saying that putting the test method into `test_base.py` caused int32 issues? Which test class did you put the method into? I'm not saying you need to put the tests in there. But I'd like to know where there might be int32 issues.  Thanks!

Reply:
It refers to this conversation (https://github.com/scipy/scipy/pull/21412#discussion_r1765053922) we had earlier in the other PR, when some CI tests were failing so we moved the _broadcast_to tests to test_csr.py. I assumed that might've been because 1D CSR used sparse arrays instead of matrices. I can move the tests to test_base.py to see if there are any failures for CSC too.